Add type annotations to `utils/shape.py`

eerovaher opened the pull request:
### Description

~The first commit only adds type annotations, the second commit also makes the `NDArrayShapeMethods` signatures explicit.~ There are a couple methods I did not annotate because they looked too complicated. I don't think that's a problem. The more code is annotated, the better type checkers can analyze it, so there's no need to get stuck on one or two methods early on.

`ShapedLikeNDArray` is a base class for many classes in `astropy`, so it would be good to get reviews from different sub-package maintainers who have expressed interest in type annotations.

UPDATE: Initially I edited the `NDArrayShapeMethods` method signatures in a separate commit because if those changes would have been controversial it would have been simple for me to undo them by dropping the commit. Based on initial reviews the explicit signatures are not a problem, so squashed the commits manually.

UPDATE 2: I have decided to update `NDArrayShapeMethods` method signatures in a separate commit (possibly separate pull request) after all because we should have tests that ensure the signatures match the corresponding `np.ndarray` method signatures.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

nstarman commented on astropy/utils/shapes.py:
```diff
@@ -385,7 +410,7 @@ def check_broadcast(*shapes):
 
 
-def unbroadcast(array):
+def unbroadcast(array: NDArray) -> NDArray:
```
NDArray can be parametrized by the dtype, which is preserved in this function.

```
def unbroadcast(array: NDArray[DT]) -> NDArray[DT]:
```

eerovaher replied:
That sounds simple enough.

nstarman commented on astropy/utils/shapes.py:
```diff
@@ -333,11 +352,17 @@ def __array_function__(self, function, types, args, kwargs):
 
 
-def check_broadcast(*shapes):
+def check_broadcast(*shapes: tuple[int, ...]) -> tuple[int, ...]:
```
it might be useful to have a `Shape` type alias defined here or in a `typing` module

eerovaher replied:
`tuple[int, ...]` is a very simple annotation and I don't think we'd gain much in readability by introducing an alias for it. However, if `numpy.typing` already has a type for shapes then it might be a good idea to use that. I am not aware of such a type, but I'm not as familiar with `numpy.typing` as I'd like to be.

mhvk replied:
See https://github.com/numpy/numpy/blob/2d2ff2501ad8a558c9ea402bf7b1a48e3f676fed/numpy/_typing/_shape.py#L4

May be worth checking with numpy to what extent these definitions are meant to be private...

eerovaher replied:
The shorter and simpler a type is the less we gain by having an alias for it. `tuple[int, ...]` is both short and simple, so using the private `numpy.typing._Shape` doesn't seem to be worth it. `SupportsIndex | Sequence[SupportsIndex]` is still simple, but it is more verbose, so using an alias is more appealing, but `_ShapeLike` is private too. In practice `int | tuple[int, ...]` (or possibly `int | Sequence[int]`) should be a good enough substitute for the actual shape-like type in most cases.

mhvk replied:
Here, I feel we should definitely not let perfect be the enemy of good enough. `tuple[int, ...]` is very clear (and plain `int` is not allowed here). I guess in the end it matters if type checking software is going to have a problem with someone calling `check_broadcast((np.int64(1),), (10,))` or `check_broadcast([1], [10])` (both of which work fine). So, in that sense `Sequence[SupportsIndex]` would be more strictly correct (but arguably less clear for the human reader...)

nstarman commented on astropy/utils/shapes.py:
```diff
@@ -204,28 +223,28 @@ def _apply(method, *args, **kwargs):
-    def size(self):
+    def size(self) -> int:
         """The size of the object, as calculated from its shape."""
         size = 1
         for sh in self.shape:
             size *= sh
         return size
```
Unrelated to this PR. This code can be optimized!

mhvk replied:
There was discussion about it at the time, and timing experiments suggested this was close to optimal already (certainly, `np.prod(self.shape)` is far slower).

astrofrog replied:
``math.prod`` would be faster and simpler than the current code:

```python
In [1]: from math import prod

In [2]: import numpy as np

In [3]: def size(shape):
   ...:     size = 1
   ...:     for sh in shape:
   ...:         size *= sh
   ...:     return size
   ...: 

In [4]: shape = (10, 20, 22)

In [5]: %timeit np.prod(shape)
2.17 µs ± 12.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [6]: %timeit prod(shape)
57.7 ns ± 1.07 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit size(shape)
81.6 ns ± 0.289 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

astrofrog replied:
Not that 30ns is going to matter much! But at least the code would be simpler.

mhvk replied:
Yes, good to know about `math.prod`. It was introduced in 3.8, so no surprise it wasn't used before. And we can safely use it now.

neutrinoceros replied:
Definitely, and I didn't mean to suggest otherwise. I just wanted to point to a similar discussion I had some time ago :)

mhvk commented on astropy/utils/shapes.py:
```diff
@@ -138,7 +149,13 @@ def squeeze(self, *args, **kwargs):
         return self._apply("squeeze", *args, **kwargs)
 
-    def take(self, indices, axis=None, out=None, mode="raise"):
+    def take(
```
In principle, these methods are meant to be similar to `ndarray`, but they look different from what numpy does: https://github.com/numpy/numpy/blob/2d2ff2501ad8a558c9ea402bf7b1a48e3f676fed/numpy/__init__.pyi#L1673-L1696

Note that I mention this mostly in case it matters; for me the numpy stuff is essentially ununderstandable, while this is quite clear, so from that point of view I prefer this!

eerovaher replied:
I did not look at any type stubs, the annotations that I added to these methods are my transcriptions of docstrings. But because you brought up the type stubs it might be good to comment on at least one of them.

For `take()` `numpy` defines three overloads. The third one is for the case where `out` is not `None`, but our method raises a `NotImplementedError` in that case, so that overload is not applicable for us. The remaining two distinguish calls that produce scalar values from calls that produce array values. For automatic type checking this distinction can be useful, but for helping IDEs perform tab-completion it is less important.

mhvk replied:
Ah, thanks. For our situation, the scalar case is also irrelevant, since we still return the same class. 

I'm very happy to just aim for ease-of-reading and tab completion in this first iteration!

mhvk commented on astropy/utils/shapes.py:
```diff
@@ -85,32 +87,32 @@ def reshape(self, *args, **kwargs):
+        return self._apply("flatten", order=order)
 
-    def transpose(self, *args, **kwargs):
+    def transpose(self, *axes: int | tuple[int, ...] | None) -> Self:
```
numpy itself does a `from typing import SupportsIndex`; see https://github.com/numpy/numpy/blob/2d2ff2501ad8a558c9ea402bf7b1a48e3f676fed/numpy/__init__.pyi#L1574-L1577

eerovaher replied:
`int` is a subtype of `SupportsIndex`, so in principle using `int` means our types are stricter than the `numpy` ones. However, the use of instances of `SupportsIndex` that are not `int` instances is very rare, so I don't consider this to make a difference in practice. Furthermore, I expect that there are many more `astropy` users who know what `int` is that there are users who know what `SupportsIndex` is.

mhvk reviewed (commented):
The first commit certainly looks good.

The second goes somewhat against my original intent, which was to leave it to subclasses to decide exactly how the various methods may behave and thus just pass on `*args, **kwargs`. That said, since the whole point is for those methods to be as close as possible to those of `ndarray`, I think it is fine to expand the signatures. The main question then is to what extent one can reuse the actual `ndarray` signatures...

eerovaher commented:
I introduced a type variable to specify that `unbroadcast()` does not change the array `dtype` and in the `NDArrayShapeMethods` method parameter annotations I replaced `int | tuple[int, ...]` with `int | Sequence[int]`. The latter is a supertype of the former and is still easily readable. The additions to `docs/nitpick-exceptions` are to allow the documentation build to succeed without warnings.

eerovaher commented on astropy/utils/shapes.py:
```diff
@@ -25,6 +28,16 @@
+    from numpy.typing import ArrayLike, NDArray
+    from typing_extensions import Self
+
+    DT = TypeVar("DT", bound=np.generic)
```
Is there a better way of writing this down? I know about [`DTypeLike`](https://numpy.org/doc/stable/reference/typing.html#dtypelike), but that's for types that can be converted to a `dtype`, not for `dtype`s themselves.

mhvk reviewed (approved):
I'm happy with this, though maybe slightly on the fence on defining a `DT` here, just for a very small benefit - it feels like something that should be done via numpy.

Approving since that's only a nitpick, but not merging so that those more invested in typing can comment further.

neutrinoceros reviewed (changes requested):
I think I spotted a couple mistakes regarding positional catch-alls.

A general concern I have here is that we don't have means to systematically check that signatures stay in sync with numpy's equivalent functions (`*args, **kwargs` are admittedly not as informative, but guarantee that we stay compatible without checking).
If `NDArraySHapeMethods` was derived from `ndarray`, we could use `typing.override` for this (I'm not suggesting to add this relationship just for the sake of this check). Maybe we could adapt my test class from https://github.com/astropy/astropy/pull/15710 ?

eerovaher commented on astropy/utils/shapes.py:
```diff
@@ -57,14 +70,16 @@ class NDArrayShapeMethods:
 
-    def reshape(self, *args, **kwargs):
+    def reshape(
+        self, *newshape: int | Sequence[int], order: Literal["C", "F", "A"] = "C"
```
The suggested code would break at least https://github.com/astropy/astropy/blob/16786add79acbd89743b006d5eccf7d270089047/astropy/coordinates/tests/test_erfa_astrom.py#L66

neutrinoceros replied:
You're right. I think I got confused while reading numpy docs. `order` is in fact *not* allowed a positional.

eerovaher commented on astropy/utils/shapes.py:
```diff
@@ -74,35 +89,35 @@ def reshape(self, *args, **kwargs):
+        return self._apply("flatten", order=order)
 
-    def transpose(self, *args, **kwargs):
+    def transpose(self, *axes: int | Sequence[int] | None) -> Self:
```
Updating `transpose()` like you suggest would break at least https://github.com/astropy/astropy/blob/16786add79acbd89743b006d5eccf7d270089047/astropy/coordinates/tests/test_representation.py#L1606

eerovaher replied:
Admittedly the annotation I wrote is a bit sloppy, a more correct one would be
```python
@overload
def transpose(self, axes: None | Sequence[int], /) -> Self: ...
@overload
def transpose(self, *axes: int) -> Self: ...
```

neutrinoceros replied:
oh right, because `axes` cannot be passed as keyword.

eerovaher commented on astropy/utils/shapes.py:
```diff
@@ -204,28 +225,28 @@ def _apply(method, *args, **kwargs):
         return self.shape[0]
 
-    def __bool__(self):
+    def __bool__(self) -> bool:
```
`bool` is not the only valid return type for `__bool__()`, it could also be `Literal[True]` or `Literal[False]`.

eerovaher commented on astropy/utils/shapes.py:
```diff
@@ -385,7 +412,7 @@ def check_broadcast(*shapes):
 
 
-def unbroadcast(array):
+def unbroadcast(array: NDArray[DT]) -> NDArray[DT]:
```
At runtime `DT` does not get used at all, and when an IDE is displaying the function signature it doesn't really make any difference where it is defined.

neutrinoceros replied:
yeah, it really just makes a tiny difference for human readers, but it's really a nit, so feel free to disregard

nstarman commented:
> A general concern I have here is that we don't have means to systematically check that signatures stay in sync with numpy's equivalent functions

There should be a check on the signatures! Either in this PR or in a followup. What do you think @eerovaher?

eerovaher commented:
I agree it would be best to have tests to ensure the `NDArrayShapeMethods` method signatures stay consistent with the corresponding `np.ndarray` signatures. I will look into that, but I don't expect to have that ready today. I have removed the signature updates from this pull request. I'll leave it to the reviewers to decide if this should be merged as is or if it would be better to wait for the `NDArrayShapeMethods` signature updates. If this is still open when I am done with the signature changes then I will push them here. If not then I'll open a separate pull request. Either way is fine with me.

mhvk reviewed (approved):
OK, let's get this piece in, and deal with the ndarray-like signatures separately.

Thanks, @eerovaher!

eerovaher commented:
I had a look at writing unit tests for ensuring the `NDArrayShapeMethods` method signatures stay consistent with the corresponding `ndarray` signatures, but [`inspect.signature()`](https://docs.python.org/3/library/inspect.html#inspect.signature) can't figure out anything about the `ndarray` methods. We could check for consistency with a type checker, but we are not using one yet, and the annotations we are currently adding are much less rigorous than the `numpy` annotations, which a type checker wouldn't like. In the short term it seems best to leave the `NDArrayShapeMethods` methods as they are and to add annotations where it is simpler.
